adapter/src/debug_session.rs

```diff
@@ -1103,6 +1133,85 @@ impl DebugSession {
         Ok(())
     }
 
+    fn handle_disassemble(&mut self, args: DisassembleArguments) -> Result<DisassembleResponseBody, Error> {
+        fn invalid_instruction() -> DisassembledInstruction {
+            DisassembledInstruction {
+                address: "".into(),
+                instruction: "<invalid>".into(),
+                ..Default::default()
+            }
+        }
+
+        let base_addr = parse_int::parse::<u64>(&args.memory_reference)?;
+        let base_addr = match args.offset {
+            Some(offset) => base_addr.wrapping_add(offset as u64),
+            None => base_addr,
+        };
+        let instruction_offset = args.instruction_offset.unwrap_or(0);
+        if args.instruction_count < 0 {
+            bail!("Invalid instruction_count");
+        }
+        let instruction_count = args.instruction_count as usize;
+
+        let mut result = if instruction_offset >= 0 {
+            let start_addr = SBAddress::from_load_address(base_addr, &self.target);
+            let instructions =
+                self.target.read_instructions(&start_addr, (instruction_offset + args.instruction_count) as u32, &self.assembly_flavor);
+
+            let mut dis_instructions = Vec::new();
+            for instr in instructions.iter().skip(instruction_offset as usize) {
+                dis_instructions.push(disassembly::sbinstr_to_disinstr(&instr, &self.target, true));
+            }
+            dis_instructions
+        } else {
+            let bytes_per_instruction = disassembly::max_instruction_bytes(&self.target);
+            let offset_bytes = -instruction_offset * bytes_per_instruction as i64;
+            let start_addr = base_addr.wrapping_sub(offset_bytes as u64);
+            let mut disassemble_bytes = instruction_count * bytes_per_instruction as usize;
+
+            let mut dis_instructions = Vec::new();
+
+            let expected_index = -instruction_offset as usize;
+
+            // we make sure to extend disassemble_bytes to ensure that base_addr
+            // is always included
+            if start_addr + ( disassemble_bytes as u64 ) < base_addr {
+              disassemble_bytes = ( base_addr - start_addr + bytes_per_instruction ) as usize;
+            }
+
+            for shuffle_count in 0..bytes_per_instruction {
+              dis_instructions = disassembly::disassemble_byte_range(start_addr - shuffle_count, disassemble_bytes, &self.target.process(), &self.assembly_flavor)?;
+              if let Some(index) = dis_instructions.iter().position( |i| parse_int::parse::<u64>(&i.address).unwrap_or(0) == base_addr ) {
```

Comment:
Sure, I'll take a look at this as a cleanup pass.

Reply:
Done. Refactored to return vec<SBInstruction> (rather than SBInstructionList) and now the two branches are more similar. Added some more commentary too.